### PR TITLE
[DependencyInjection][Routing] Access environment in PHP config

### PR DESCRIPTION
**Pull request by Nyholm**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | 
| Tickets       | 
| License       | MIT
| Doc PR        |

This will allow me to write config like: 

```php
use Symfony\Config\FrameworkConfig;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function (FrameworkConfig $framework, ContainerConfigurator $container) {
    if ('test' === $container->env()) {
        $framework->test(true);
        $framework->session()->storageFactoryId('session.storage.mock_file');
    }
};
```

This PR will also revert parts of #40214. It is much simpler to maintain and write PHP config without `->when()`. Instead we add `ContainerConfigurator::env(): ?string` and `RoutingConfigurator::env(): ?string`. 

```php

use App\Controller\BlogController;
use Symfony\Component\Routing\Loader\Configurator\RoutingConfigurator;

return function (RoutingConfigurator $routes) {
    if ($routes->env() === 'dev') {
        $routes->add('debug-stats', '/blog-debyg')
            ->controller([BlogController::class, 'debug'])
        ;
    }

    $routes->add('blog_list', '/blog')
        ->controller([BlogController::class, 'list'])
    ;
};
```

### CODE REVIEW

**Review by yceruto — DISMISSED** [review]
Would it make sense for you to have a `->when('test')` method as well? Like we recently did for `ContainerConfigurator`.

**Comment by Nyholm** [comment]
I will actually work on remove the `when()` method. It is not really needed in PHP, right? 

I like it in yaml and XML because it allows me to write simple logic, however, in PHP I do not need a feature to "help me with simple logic".

**Comment by yceruto** [comment]
I was thinking about this config based on your example:
```php
use Symfony\Config\FrameworkConfig;

return static function (FrameworkConfig $framework)
{
    $framework->when('test')
        ->test(true)
        ->session()->storageFactoryId('session.storage.mock_file')
    ;
};
```
Not sure if possible though (I didn't check if the generated Config is fluent).

**Comment by Nyholm** [comment]
Yes, That is the alternative way of doing things. The implementation is more difficult and it is less obvious what is happening.

**Comment by Nyholm** [comment]
Hm.. So we are loading callbacks on 4 places. 

1) DI: Microkernel trait
2) DI: PhpFileLoader
3) Router: Microkernel trait
4) Router: PhpFileLoader

For 2, we have a fancy resolver of the parameters. So order does not matter, but parameter name does. For the others, order does matter but not parameter name. 

The fancy resolver is needed for the ConfigBuilders. I would like to use fancy resolving for all of these, but either I write (almost) the same code on 3 more places or I try to do something complex...

**Comment by Nyholm** [comment]
So, instead of doing fancy argument resolving on all places. I added `ContainerConfigurator::env(): ?string` and `RoutingConfigurator::env(): ?string`. 

I've updated the description. This PR is ready for a review.

File: src/Symfony/Component/DependencyInjection/Loader/Configurator/ContainerConfigurator.php [hunk 1]
```diff
@@ -74,20 +74,11 @@ final public function services(): ServicesConfigurator
     }
 
     /**
-     * @return static
+     * Get the current environment to be able to write conditional configuration.
      */
-    final public function when(string $env): self
+    public function env(): ?string
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/Configurator/ContainerConfigurator.php]: should be final

File: src/Symfony/Component/DependencyInjection/Loader/PhpFileLoader.php [hunk 2]
```diff
@@ -117,6 +117,13 @@ private function executeCallback(callable $callback, ContainerConfigurator $cont
                 case self::class:
                     $arguments[] = $this;
                     break;
+                case 'string':
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/PhpFileLoader.php]: do we want to keep that case, or should we remove support for this argument in favor of the env() method, in order to have only one way to do this env-related check?

Comment (Nyholm) [reply src/Symfony/Component/DependencyInjection/Loader/PhpFileLoader.php]: I very much like this. To be able to write this is super cool: 

```php
return static function (FrameworkConfig $framework, string $env) {
```

However, the easier concept to understand is to do `ContainerConfigurator::env()` as you suggest. The `string $env` will only work "sometimes"...

```php
return static function (FrameworkConfig $framework, ContainerConfigurator $container) {
    $env = $container->env();
```

I'll remove this.

File: src/Symfony/Component/DependencyInjection/CHANGELOG.md [hunk 3]
```diff
@@ -10,11 +10,13 @@ CHANGELOG
  * Add `#[AsTaggedItem]` attribute for defining the index and priority of classes found in tagged iterators/locators
  * Add autoconfigurable attributes
  * Add support for autowiring tagged iterators and locators via attributes on PHP 8
- * Add support for per-env configuration in loaders
+ * Add support for per-env configuration in XML and Yaml loaders
  * Add `ContainerBuilder::willBeAvailable()` to help with conditional configuration
  * Add support an integer return value for default_index_method
  * Add `env()` and `EnvConfigurator` in the PHP-DSL
  * Add support for `ConfigBuilder` in the `PhpFileLoader`
+ * Add support for parameter `string $env` in PHP config closures
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/CHANGELOG.md]: should be removed

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/DependencyInjection/CHANGELOG.md (hunk 3 above)
Comment (Nyholm) [reply src/Symfony/Component/DependencyInjection/CHANGELOG.md]: Oops, sorry. I've updated the PR

**Review by yceruto — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @Nyholm.

**Comment by Nyholm** [comment]
Woho. Let's start write php config

**Pull request merged**